### ISSUE TITLE
error authorizing context: authorization token required

### ISSUE THREAD

**Issue by ChrisB85** [zebox/registry-admin#20]
Hi,
I'm trying to run the app in localhost (for testing purposes for now) using files from _examples/token_auth. I am able to login, but when I hit "Sync" button i see some error in log:
`docker-registry-registry-1        | time="2023-08-09T18:24:10.564085228Z" level=warning msg="error authorizing context: authorization token required" go.version=go1.19.9 http.request.host="registry:5000" http.request.id=10dd6015-f1da-4b4f-97af-62b008efad78 http.request.method=GET http.request.remoteaddr="172.19.0.2:38366" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1"`

What can be the cause of this?

docker-compose file:
```
version: '2.1'
services:
  registry-admin:
    restart: unless-stopped
    image: zebox/registry-admin:master
    ports:
      - 80:80
    environment:
      - RA_CONFIG_FILE=/app/config/token-ra-config.yml
    volumes:
      - ./certs:/app/certs
      - ./config:/app/config
      - ./data:/app/data

  registry:
    restart: unless-stopped
    image: registry:2
    ports:
      - 50554:5000
    environment:
      - REGISTRY_AUTH_TOKEN_REALM=http://localhost/api/v1/registry/auth # <- !!! Change this for real hostname of RegistryAdmin and exposed port which accessible for docker clients
    volumes:
      - ./data:/var/lib/registry
      - ./certs:/certs
      - ./config/registry-config.yml:/etc/docker/registry/config.yml
    depends_on:
      - registry-admin
```

registry-config.yml:
```
version: 0.1

log:
  accesslog:
    disabled: false
  level: debug
  formatter: text
  fields:
    service: registry

storage:
  filesystem:
    rootdirectory: /var/lib/registry
    maxthreads: 100
  delete:
    enabled: true

http:
  addr: ":5000"
  net: tcp
  tls:
    certificate: /certs/cert.crt
    key: /certs/cert.key

auth:
  token:
    realm: http://localhost/api/v1/registry/auth  # <- external ip accessible for clients from outside of container
    service: container_registry
    issuer: registry_token_issuer
    rootcertbundle: /certs/cert.crt

notifications:
  events:
    includereferences: true
  endpoints:
    - name: ra-listener
      disabled: false
      url: http://registry-admin/api/v1/registry/events
      headers:
        Authorization: [Basic YWRtaW46c3VwZXItc2VjcmV0] # 'admin:super-secret' base64 encode string
      timeout: 1s
      threshold: 5
      backoff: 3s
      ignoredmediatypes:
        - application/octet-stream
      ignore:
        mediatypes:
          - application/octet-stream
```

 token-ra-config.yml:
```
hostname: localhost

registry:
  host: https://registry
  port: 5000
  auth_type: token
  issuer: registry_token_issuer
  service: container_registry
  certs:
    path: /app/certs
    key: /app/certs/cert.key
    public_key: /app/certs/cert.pub
    ca_root: /app/certs/cert.crt
    ip: 127.0.0.1 # <- paste a real IP of docker host which publish the container
    fqdns: [registry, demo.host.local, localhost, registry.host.local]

store:
  type: embed
  admin_password: "super-secret"
  embed:
    path: /app/data/store.db
```

**Comment by zebox** [comment]
Hi @ChrisB85!

I suppose mistake in this
`realm: http://localhost/api/v1/registry/auth`

You should change `localhost` to hostname  which defined for the registry-admin container. In `docker-compose.yml` you can define `hostname` derective or use service name as hostname. 

In your case the registry-admin service has name `registry-admin` and this name available with mDNS inside containers stack network.

Target realm for token, in registry config file, should has hostname `registry-admin`:

`realm: http://registry-admin/api/v1/registry/auth`

**Comment by ChrisB85** [comment]
@zebox  Thak you for reply.

I made change you suggested, but unfortunately the `authorization token required` error still appears.

**Comment by zebox** [comment]
I will try reproduce this issue in my environment in nearest time. Until you can try use works step-by-step example from my [article](https://habr.com/ru/articles/709988), but it on russian language.

**Comment by zebox** [comment]
I checked config example and all work properly at me. 
But has some clarification:

1. Environment vars override config setting and you should change to `registry-admin` host name either in the env var `REGISTRY_AUTH_TOKEN_REALM` or comment out it (or remove) and define `realm` host name in registry config file only.
```yml
environment:
      - REGISTRY_AUTH_TOKEN_REALM=http://registry-admin/api/v1/registry/auth # <- !!! Change this for real host name of RegistryAdmin and exposed port which accessible for docker clients
```
2. For testing in local environment I recommend disable TLS for access to registry. It help you to avoid difficult with certificate authority trust issue. Otherwise you need configure self-signed certificate to trusted CA in registry container and insecure TLS  for a docker engine at client host. I will make step-by-step guide for this later.

This works example for your testing:

docker-compose.yml 
```yml
version: '2.1'
services:
  registry-admin:
    hostname: registry-admin
    restart: unless-stopped
    image: zebox/registry-admin:master
    ports:
      - 8080:80
    environment:
      - RA_CONFIG_FILE=/app/config/token-ra-config.yml
    volumes:
      - ./certs:/app/certs
      - ./config:/app/config
      - ./data:/app/data

  registry:
    restart: unless-stopped
    image: registry:2
    ports:
      - 50554:5000
    volumes:
      - ./data:/var/lib/registry
      - ./certs:/certs
      - ./config/registry-config.yml:/etc/docker/registry/config.yml
    depends_on:
      - registry-admin
```
registry-config.yml

```yml
version: 0.1

log:
  accesslog:
    disabled: false
  level: debug
  formatter: text
  fields:
    service: registry

storage:
  filesystem:
    rootdirectory: /var/lib/registry
    maxthreads: 100
  delete:
    enabled: true

http:
  addr: ":5000"
  net: tcp
  # disable TLS access for testing in local env
  #tls:
    #certificate: /certs/cert.crt
    #key: /certs/cert.key

auth:
  token:
    realm: http://registry-admin/api/v1/registry/auth  # <- external ip accessible for clients from outside of container
    service: container_registry
    issuer: registry_token_issuer
    rootcertbundle: /certs/cert.crt

notifications:
  events:
    includereferences: true
  endpoints:
    - name: ra-listener
      disabled: false
      url: http://registry-admin/api/v1/registry/events
      headers:
        Authorization: [Basic YWRtaW46c3VwZXItc2VjcmV0] # 'admin:super-secret' base64 encode string
      timeout: 1s
      threshold: 5
      backoff: 3s
      ignoredmediatypes:
        - application/octet-stream
      ignore:
        mediatypes:
          - application/octet-stream
```
token-ra-config.yml:
```yml
hostname: 127.0.0.1

registry:
  host: http://registry
  port: 5000
  auth_type: token
  issuer: registry_token_issuer
  service: container_registry
  certs:
    path: /app/certs
    key: /app/certs/cert.key
    public_key: /app/certs/cert.pub
    ca_root: /app/certs/cert.crt
    ip: {YOUR-OUTSIDE-CONTAINER-IP} # <- paste a real IP of docker host which publish the container
    fqdns: [demo.host.local, localhost, registry.host.local]  # <- paste your fqdns

store:
  type: embed
  admin_password: "super-secret" # you should delete this string after database created
  embed:
    path: /app/data/store.db
```

**Comment by ChrisB85** [comment]
I have copied the content of examples, deleted certs, store.db, but still error persists.  Here's the log from Sync operation:
```
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:41:56 +0000] "GET /auth/user HTTP/1.1" 200 158 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:41:56 +0000] "GET /api/v1/registry/catalog?filter=%7B%7D&range=%5B0%2C24%5D&sort=%5B%22repository_name%22%2C%22ASC%22%5D& HTTP/1.1" 200 24 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:48:08 +0000] "GET /auth/user HTTP/1.1" 200 158 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:48:08 +0000] "GET /api/v1/registry/catalog?filter=%7B%7D&range=%5B0%2C24%5D&sort=%5B%22repository_name%22%2C%22ASC%22%5D& HTTP/1.1" 200 24 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:48:08 +0000] "GET /auth/user HTTP/1.1" 200 158 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:48:08 +0000] "GET /auth/user HTTP/1.1" 200 158 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry-admin_1  | 192.168.0.100 - - [10/Aug/2023:19:48:10 +0000] "GET /api/v1/registry/sync?filter=%7B%7D&range=%5B0%2C9%5D&sort=%5B%22id%22%2C%22DESC%22%5D& HTTP/1.1" 200 48 "http://192.168.0.2:8066/registry/catalog" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36"
registry_1        | time="2023-08-10T19:48:10.010707458Z" level=debug msg="authorizing request" go.version=go1.19.9 http.request.host="registry:5000" http.request.id=b67384c1-1360-4bbe-8524-06df2d789e6f http.request.method=GET http.request.remoteaddr="192.168.16.2:49750" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1"
registry_1        | time="2023-08-10T19:48:10.01250272Z" level=warning msg="error authorizing context: authorization token required" go.version=go1.19.9 http.request.host="registry:5000" http.request.id=b67384c1-1360-4bbe-8524-06df2d789e6f http.request.method=GET http.request.remoteaddr="192.168.16.2:49750" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1"
registry_1        | 192.168.16.2 - - [10/Aug/2023:19:48:10 +0000] "GET /v2/_catalog?n=50&last= HTTP/1.1" 401 145 "" "Go-http-client/1.1"
registry_1        | time="2023-08-10T19:48:10.028735918Z" level=debug msg="authorizing request" go.version=go1.19.9 http.request.host="registry:5000" http.request.id=ff69c809-e94f-4103-9f5c-8a450af3b984 http.request.method=GET http.request.remoteaddr="192.168.16.2:54994" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1"
registry_1        | time="2023-08-10T19:48:10.02964609Z" level=info msg="authorized request" go.version=go1.19.9 http.request.host="registry:5000" http.request.id=ff69c809-e94f-4103-9f5c-8a450af3b984 http.request.method=GET http.request.remoteaddr="192.168.16.2:54994" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1"
registry_1        | time="2023-08-10T19:48:10.029999242Z" level=debug msg="filesystem.Walk("/docker/registry/v2/repositories")" auth.user.name= go.version=go1.19.9 http.request.host="registry:5000" http.request.id=ff69c809-e94f-4103-9f5c-8a450af3b984 http.request.method=GET http.request.remoteaddr="192.168.16.2:54994" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1" trace.duration=67.668µs trace.file="github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Walk" trace.id=6015c66b-85d3-4aec-9a58-aa0497853b81 trace.line=232
registry_1        | time="2023-08-10T19:48:10.030321688Z" level=info msg="response completed" go.version=go1.19.9 http.request.host="registry:5000" http.request.id=ff69c809-e94f-4103-9f5c-8a450af3b984 http.request.method=GET http.request.remoteaddr="192.168.16.2:54994" http.request.uri="/v2/_catalog?n=50&last=" http.request.useragent="Go-http-client/1.1" http.response.contenttype="application/json; charset=utf-8" http.response.duration=3.815358ms http.response.status=200 http.response.written=20
registry_1        | 192.168.16.2 - - [10/Aug/2023:19:48:10 +0000] "GET /v2/_catalog?n=50&last= HTTP/1.1" 200 20 "" "Go-http-client/1.1"
registry-admin_1  | 2023/08/10 19:48:10.030 [ERROR] garbage collector skip because sync required start first
registry-admin_1  | 2023/08/10 19:48:10.030 [ERROR] garbage collector skip because sync required start first
```
I ran it in WSL and my homelab server (with Ubuntu on board). The above result is the same for both.

The only doubt I have is about:
`ip: {YOUR-OUTSIDE-CONTAINER-IP} # <- paste a real IP of docker host which publish the container`
I tried entering phisical network card IP (192.168.0.100 and 192.168.0.2) and in WSL also IP which is returned by `ip -4 addr|grep inet` command (172.21.16.242). No luck.

**Comment by zebox** [comment]
It's very strange(. I use WSL env too in my dev computer. But I tested it on Linux machine and all worked properly.

I will record step by step video with example config for you.

**Comment by ChrisB85** [comment]
Repositiry is empty, but when I try to push some image I get:
```
The push refers to repository [localhost/someimage]
5f70bf18a086: Preparing
865c2d40902c: Preparing
d6df4d8f1d2b: Preparing
814b91d8ce2a: Preparing
243673d2c35b: Preparing
e0e5c14f683a: Waiting
3e8300fe133b: Waiting                                                                                                                                                   cd6c2464dc51: Waiting
f7f303b50df0: Waiting
d5fcc5cffc99: Waiting                                                                                                                                                   1fdbf5f06e1e: Waiting
eb5e1abd9327: Waiting                                                                                                                                                   1059c10ff87a: Waiting
3cdaf4f3899c: Waiting
57d046864aa6: Waiting
3eb0486809d0: Waiting
1efc5401b6f1: Waiting
b31fe8530467: Waiting
c151effcd197: Waiting
0b0b8e9d2e04: Waiting
a9ed1f92fa62: Waiting
6e28a572644e: Waiting
67b5ce3064ab: Waiting
88cd9b949e2e: Waiting
e2ef8a51359d: Waiting
error parsing HTTP 405 response body: unexpected end of JSON input: ""
```

**Comment by zebox** [comment]
You can't push images until logged in to the registry. Also, I saw that registry URL missed port (`50554` defined in compose file) which expose from container.

I excuse for confusing, but `realm` URL should be accessed from container outside.

I will prepare tutorial video with detailed description soon.

**Comment by ChrisB85** [comment]
Yes, I was missing repository port, now I finally got it working 😄 Thank you for your help.
It seems that `error authorizing context: authorization token required` appeard becouse repository was empty. When I pushed first image to it the sync is sucessfull.
So everything is working now in localhost environment, but in my homelab server I got `unauthorized: authentication required` in the middle of push process. I will create another ticket for that.
